modules/client/src/main/java/org/apache/ignite/internal/client/table/ClientPartitionManager.java

```diff
@@ -0,0 +1,198 @@
+
+                    if (in.tryUnpackNil()) {
+                        return null;
+                    }
+
+                    int size = in.unpackInt();
+
+                    Map<Partition, ClusterNode> res = new HashMap<>(size);
+
+                    for (int i = 0; i < size; i++) {
+                        int partition = in.unpackInt();
+                        res.put(new ClientHashPartition(partition), unpackClusterNode(r));
+                    }
+
+                    return res;
+                });
+    }
+
+    @Override
+    public <K> CompletableFuture<Partition> partitionAsync(K key, Mapper<K> mapper) {
+        return getPartition(getPartitionAwarenessProvider(null, mapper, key),
+                (schema, w) -> ClientRecordSerializer.writeRecRaw(key, mapper, schema, w.out(), TuplePart.KEY)
+        );
+    }
+
+    @Override
+    public CompletableFuture<Partition> partitionAsync(Tuple key) {
+        return getPartition(getPartitionAwarenessProvider(null, key),
+                (schema, w) -> ClientTupleSerializer.writeTupleRaw(key, schema, w, true)
+        );
+    }
+
+    @SuppressWarnings("resource")
+    private CompletableFuture<Partition> getPartition(
+            PartitionAwarenessProvider partitionAwarenessProvider,
+            BiConsumer<ClientSchema, PayloadOutputChannel> keyPack
+    ) {
+        return tbl.getPartitionAssignment()
+                .thenCompose(partitions -> tbl.getLatestSchema().thenApply(schema -> {
+                    Integer hash = partitionAwarenessProvider.getObjectHashCode(schema);
+
+                    return Math.abs(hash % partitions.size());
+                }))
+                .exceptionally(t -> -1)
+                .thenCompose(partition -> {
+                    if (partition == -1) {
+                        return tbl.getLatestSchema()
+                                .thenCompose(schema ->
+                                        tbl.channel().serviceAsync(ClientOp.KEY_PARTITION_PRIMARY_GET,
+                                                w -> {
+                                                    w.out().packInt(tbl.tableId());
+                                                    keyPack.accept(schema, w);
+                                                },
+                                                r -> r.in().unpackInt()
+                                        ));
+                    }
+
+                    return completedFuture(partition);
+                }).thenApply(ClientHashPartition::new);
+    }
+
+    private static @Nullable ClusterNode unpackClusterNode(PayloadInputChannel r) {
+        ClientMessageUnpacker in = r.in();
+        if (in.tryUnpackNil()) {
+            return null;
+        }
+
+        return new ClientClusterNode(
+                in.unpackString(),
+                in.unpackString(),
+                new NetworkAddress(in.unpackString(), in.unpackInt()));
+    }
+
+    public static class ClientHashPartition implements Partition {
```

Comment:
This is a copy of `HashPartition`. We should reuse implementations across embedded and client APIs when possible. Let's move `HashPartition` to the `core` module.